docs/gitbook/embedding/word2vec.md

```diff
@@ -0,0 +1,399 @@
+because it saves memory during training.
+
+# Create sub-sampling table
+
+Sub-sampling table is stored a sub-sampling probability per word.
+
+The sub-sampling probability of word $$w_i$$ is computed by the following equation:
+
+$$
+\begin{aligned}
+f(w_i) = \sqrt{\frac{\mathrm{sample}}{freq(w_i)/\sum freq(w)}} + \frac{\mathrm{sample}}{freq(w_i)/\sum freq(w)}
+\end{aligned}
+$$
+
+During word2vec training,
+not sub-sampled words are ignored.
+It works to train fastly and to consider the imbalance the rare words and frequent words by reducing frequent words.
+The smaller `sample` value set,
+the fewer words are used during training.
+
+```sql
+set hivevar:sample=1e-4;
+
+drop table subsampling_table;
+create table subsampling_table as
+with stats as (
+  select
+    sum(freq) as numTrainWords
+  FROM
+    freq
+)
+select
+  l.wordid,
+  l.word,
+  sqrt(${sample}/(l.freq/r.numTrainWords)) + ${sample}/(l.freq/r.numTrainWords) as p
+from
+  freq l
+cross join
+  stats r
+;
+```
+
+```sql
+select * FROM subsampling_table order by p;
+```
+
+| wordid | word | p |
+|:----: | :----: |:----:|
+| 48645 | the  | 0.04013665|
+| 11245 | of   | 0.052463654|
+| 16368 | and  | 0.06555538|
+| 61938 | 00   | 0.068162076|
+| 19977 | in   | 0.071441144|
+| 83599 | 0    | 0.07528994|
+| 95017 | a    | 0.07559573|
+| 1225  | to   | 0.07953133|
+| 37062 | 0000 | 0.08779001|
+| 58246 | is   | 0.09049763|
+|  ...  | ...  |... |
+
+The first row shows that 4% of `the` are used in the documents during training.
+
+# Delete low frequency words and high frequency words from `docs_words`
+
+To reduce useless words from corpus,
+low frequency words and high frequency words are deleted.
+And, to avoid loading long document on memory, a  document is split into some sub-documents.
+
+```sql
+set hivevar:maxlength=1500;
+SET hivevar:seed=31;
+
+drop table train_docs;
+create table train_docs as
+  with docs_exploded as (
+    select
+      docid,
+      word,
+      pos % ${maxlength} as pos,
+      pos div ${maxlength} as splitid,
+      rand(${seed}) as rnd
+    from
+      docs_words LATERAL VIEW posexplode(words) t as pos, word
+  )
+select
+  l.docid,
+  -- to_ordered_list(l.word, l.pos) as words
+  to_ordered_list(r2.wordid, l.pos) as words,
+from
+  docs_exploded l
+  LEFT SEMI join freq r on (l.word = r.word)
+  join subsampling_table r2 on (l.word = r2.word)
+where
+  r2.p > l.rnd
+group by
+  l.docid, l.splitid
+;
+```
+
+If you store string word in `train_docs` table,
+please replace `to_ordered_list(r2.wordid, l.pos) as words` with  `to_ordered_list(l.word, l.pos) as words`.
+
+# Create negative sampling table
+
+Negative sampling is an approximate function of [softmax function](https://en.wikipedia.org/wiki/Softmax_function).
+Here, `negative_table` is used to store word sampling probability for negative sampling.
+`z` is a hyperparameter of noise distribution for negative sampling.
+During word2vec training,
```

Comment:
Line break is not needed. Line break after `,` is unreasonable (elsewhere as well).